compiler/rustc_typeck/src/check/generator_interior.rs

```diff
@@ -409,3 +448,170 @@ impl<'a, 'tcx> Visitor<'tcx> for ArmPatCollector<'a> {
         }
     }
 }
+
+#[derive(Default)]
+pub struct SuspendCheckData<'a, 'tcx> {
+    expr: Option<&'tcx Expr<'tcx>>,
+    source_span: Span,
+    yield_span: Span,
+    descr_pre: &'a str,
+    descr_post: &'a str,
+    plural_len: usize,
+}
+
+// Returns whether it emitted a diagnostic or not
+// Note that this fn and the proceding one are based on the code
+// for creating must_use diagnostics
+pub fn check_must_not_suspend_ty<'tcx>(
```

Comment:
The complexity of this function seems ~okay to me, but I'm still wondering if the (discussed and dismissed in the RFC) marker trait (`Suspend`?) approach would work better. The attribute could impl `!Suspend` for structs, add it as a supertrait for traits and then the one caller to `check_must_not_suspend_ty` could make the InferCtxt add an obligation for `Suspend` to exist.

This would require some extra logic to make that obligation failure a lint instead of an error, so if we don't already have something like this, we should keep doing what we're doing here. Maybe just add the above paragraph as a comment for why we are implementing it this way?

Reply:
Is the choice basically "we decided against this in the RFC"?

Reply:
it was never discussed in the RFC except for an honorable mention [here](https://github.com/rust-lang/rfcs/blob/master/text/3014-must-not-suspend-lint.md#auto-trait-vs-attribute)